### PR TITLE
Implement Copy operations using Operation<T> as LROs

### PR DESCRIPTION
**Pull request by tg-msft**
This is based off @KrzysztofCwalina's prototype as described by https://azuresdkspecs.z5.web.core.windows.net/DotNetSpec.html#dotnet-longrunning

The only changes for this PR are in the last commit.

### CODE REVIEW

File: sdk/storage/Azure.Storage.Blobs/src/Models/CopyFromUriOperation.cs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        /// </summary>
+        internal CopyFromUriOperation(
+            string copyId,
+            bool hasCompleted,
+            long? value = default,
+            Response rawResponse = default)
+        {
+            this._copyId = copyId;
+            this._hasCompleted = hasCompleted;
+            if (value != null)
```
Comment (kfarmer-msft) [inline sdk/storage/Azure.Storage.Blobs/src/Models/CopyFromUriOperation.cs]: why not just store the nullable directly????

**Review by kfarmer-msft — CHANGES_REQUESTED** [review]
Pretty concerned about the re-invention of Nullable<T>

File: sdk/storage/Azure.Storage.Blobs/src/Models/CopyFromUriOperation.cs (hunk 1 above)
Comment (kfarmer-msft) [reply sdk/storage/Azure.Storage.Blobs/src/Models/CopyFromUriOperation.cs]: Not going to block on this, but this sort of thing raises big red maintainability flags for me

**Review by kfarmer-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs [hunk 2]
```diff
@@ -664,27 +664,34 @@ protected virtual BlobBaseClient WithSnapshotImpl(string snapshot)
...
-        /// A <see cref="Response{BlobCopyInfo}"/> describing the
+        /// A <see cref="CopyFromUriOperation"/> describing the
         /// state of the copy operation.
         /// </returns>
         /// <remarks>
         /// A <see cref="StorageRequestFailedException"/> will be thrown if
         /// a failure occurs.
         /// </remarks>
-        public virtual Response<BlobCopyInfo> StartCopyFromUri(
+        public virtual CopyFromUriOperation StartCopyFromUri(
```
Comment (KrzysztofCwalina) [inline sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs]: After preview we should try to return Operation<T>

File: sdk/storage/Azure.Storage.Blobs/src/Models/CopyFromUriOperation.cs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        /// </summary>
+        private readonly string _copyId;
+
+        /// <summary>
+        /// Gets the ID of the copy operation that can be compared with the
+        /// <see cref="BlobProperties.CopyId"/> value returned from
+        /// <see cref="BlobBaseClient.GetPropertiesAsync" /> or passed to
+        /// <see cref="BlobBaseClient.AbortCopyFromUriAsync" />.
+        /// </summary>
+        public virtual string Id => this._copyId;
```
Comment (KrzysztofCwalina) [inline sdk/storage/Azure.Storage.Blobs/src/Models/CopyFromUriOperation.cs]: Why virtual?

**Review by KrzysztofCwalina — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs/src/Models/CopyFromUriOperation.cs (hunk 3 above)
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs/src/Models/CopyFromUriOperation.cs]: For mocking.

File: sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs (hunk 2 above)
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs]: Agreed - once we add `Operation<T>.Id`.

File: sdk/storage/Azure.Storage.Blobs/src/Models/CopyFromUriOperation.cs (hunk 1 above)
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs/src/Models/CopyFromUriOperation.cs]: What's the difference between `Value == null` while the status is Pending and `Value == null` when the status is Failed?  You need an extra indicator somewhere.

Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs/src/Models/CopyFromUriOperation.cs]: And to add, `HasCompleted` was originally going to be `OperationStatus` or similar but was simplified to `HasCompleted` to be as general purpose as possible.

**Pull request merged**